src/osd/PrimaryLogPG.cc

```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+struct C_gather : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  PrimaryLogPG::OpContext *ctx;
+  C_gather(PrimaryLogPG *pg_, hobject_t oid_, PrimaryLogPG::OpContext *ctx_) : pg(pg_), oid(oid_), ctx(ctx_) {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    // TODO: check errors
+    pg->lock();
+    pg->cls_gather_ops.erase(oid);
+    pg->execute_ctx(ctx);
+    pg->unlock();
+  }
+};
+
+int PrimaryLogPG::cls_gather(OpContext *ctx, std::map<std::string, bufferlist*> *src_objs, const std::string& pool,
+			     const char *cls, const char *method, bufferlist& inbl)
+{
+  OpRequestRef op = ctx->op;
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  object_locator_t oloc = m->get_object_locator();
+
+  ObjectState& obs = ctx->new_obs;
+  object_info_t& oi = obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  ObjectContextRef obc = get_object_context(soid, false);
+  C_GatherBuilder gather(cct);
+
+  CLSGatherOpRef cgop(std::make_shared<CLSGatherOp>(ctx, obc, op));
+  for (std::map<std::string, bufferlist*>::iterator it = src_objs->begin(); it != src_objs->end(); it++) {
+    std::string oid = it->first;
+    it->second = new bufferlist;
```

Comment:
Can src_objs simply be a std::map<string, bufferlist>?  I don't really think we want to do allocation here and trust that the caller will correctly free it.

Reply:
Actually, the user should probably pass in const std::set<std::string> & and internally we should construct a std::map<std::string, std::bufferlist> to be returned via std::move in cls_cxx_get_gathered_data.  That would entirely avoid reference ownership ambiguity.